up-l1/someip.adoc

```diff
@@ -0,0 +1,265 @@
+|===
+
+
+=== UPayload
+
+`UPayload` hosts the application layer data that is being sent between devices.
+
+`payload` structure, defined in link:../basics/umessage.adoc[UMessage] contains either a copy or reference to the actual payload that is to be sent.
+`UMessage.attributes` also contains link:../basics/upayloadformat.adoc[UPayloadFormat] used to give a hint of the payload format (protobuf serialized, SOME/IP format, TEXT, RAW, etc...).
+
+The *SOME/IP* specification however does not have an equivalent field for `UPayloadFormat` as it is assumed that the payload is serialized in the format that the other end knows how to deserialize (i.e. it is fixed per topic). As such, when converting between *uProtocol* and *SOME/IP*, the `UPayloadFormat` field *SHOULD* be ignored (left at the default of `UMESSAGE_TYPE_UNSPECIFIED`).
+
+
+=== uTransport
+
+Mapping of uTransport APIs to *SOME/IP* specific library APIs shall not be covered in this document given there are multiple open source libraries available for *SOME/IP*.
+
+
+== Application Layer Translation
+
+Application (or message payload) translation is the process of converting *SOME/IP-SD* subscription and discovery messages, to/from *uDiscovery* and *uSubscription* Messages.
+
+=== uSubscription
+
+The following section will elaborate only on the translation of *uSubscription* messages to/from *SOME/IP-SD* messages. Subscription state (persistent or not) is handled in the *uSubscription* services and not at the transport layer or this component.
+
+The following section we will elaborate on how Eventgroup Entry types are mapped to link:../up-l3/usubscription/v3/README.adoc[*uSubscription*] messages for the subscribe
+and unsubscribe flows per `[PRS_SOMEIPSD_00385]`.
+
+==== Common Fields
+
+<<common-field-mappings>> table below illustrates the common *SOME/IP-SD* EventGroup Entry fields that are present in for all *SOME/IP-SD* Eventgroup entry types (`SubscribeEventgroup`, `SubscribeEventGroupAck`, `SubscribeEventgroupNack`, `StopSubscribeEventGroup`).
+
+These fields are then mapped to uProtocol `UUri` attributes used in uProtocol `UMessage` for performing subscription operations.
+
+.Common Field Mappings
+[#common-field-mappings,  cols="1,2"]
+|===
+| Eventgroup Entry Field | UUri
+
+a| `Service ID` a| Set in lower 16 bits of `ue_id`
+
+a| `Instance ID` a| If instance is not the default (`0x1`), set it in upper 16 bits of `ue_id`
+
+a| `Major Version` a| `ue_version_major`
+
+a| `EventGroup ID` / `Event ID` a| `resource_id`
+
+NOTE: `1:1` mapping between `EventGroup ID` and `Event ID` is assumed. This may require specific ECU Firmware.
+
+
+|===
+
+NOTE: `UUri.authority_name` *MAY* be translated to/from IPv4 (and/or IPv6) Endpoint Option of the *SOME/IP-SD* message, although in `vsomeip` this is not available in the API (e.g. each discovered Endpoint maps to `Service ID`/`Instance ID`/`Major Version`/`Minor Version`).
```

Comment:
This is where we need to chat, I'm not sure how that would work. How do we know that service_id+instance_id maps to a specific IP address and port, is that static configuration that would be loaded into the uTransport implementation library?

Reply:
Also very curious about this point.

I don't know vsomeip all that well, but from what I could see it seems to rely on static configuration of IP addresses and I couldn't figure out reading the vsomeip header how this tied back to an IP address.

If the answer is as you wrote and there is no IP address or logical name contained within the vsomeip header, how do we come up with this?

Must we then read that same vsomeip configuration to be able to correlate service and instance IDs with the IP address?

Reply:
Some other SOME/IP implementations, require you to specify all the local and remote endpoints in their static configuration along with their service/instance/methods, etc.

For `vsomeip` it can be also static (if service discovery is disabled) or auto-populated from service discovery.
Consider you want to use {service:0x1234, instance:2, method:0x0042, major_ver:1}

- The app registers `on_available(service, instance, is_available)` callback, possibly with ANY_SERVICE, ANY_INSTANCE artguments.
- App calls `app->request_service(0x1234, ANY_INSTANCE, 1, ANY_MINOR)`
- This generates Find Service Entry SOME/IP message `[PRS_SOMEIPSD_00350]`
and all mDevice endpoints must return their 0x1234 service endponits (e.g. hostA provides instance 1, hostB provides instance 2)
- App receives `on_available(0x1234, 1, true)` and `on_available(0x1234, 2, true)` and then it can communicate with the choosen instance
- e.g. send a message to configured instance 2:
```c++
std::shared_ptr<vsomeip::message> rq = vsomeip::runtime::get()->create_request(use_tcp_);
rq->set_service(0x1234);
rq->set_instance(2);
rq->set_method(0x0042);
rq->set_interface_version(1); // major_ver
std::shared_ptr<vsomeip::payload> pl = vsomeip::runtime::get()->create_payload();
// serialize payload for method_id
rq->set_payload(pl);
app_->send(rq);
```
- vsomeip finds the discovered service/instance(2) and sends the request to `HostB` endpoint (app does not care about which host is providing it)
- You can also set `rq->set_instance(1)` and it will be sent to the other `hostA`